### PR TITLE
Fix FSDP for Models with Frozen Weights

### PR DESCRIPTION
**Pull request by Liyang90**
Fixing the issue where XLAFSDP is not freeing the full params of the frozen parameters (`requires_grad==False`) after a backward pass.

### CODE REVIEW

**Comment by JackCaoG** [comment]
any chance we can add a small test to make sure this does not regress?

**Comment by gkroiz** [comment]
Just want to clarify here, the user should freeze parameters before sharding? What difference does it make if the parameters are frozen after sharding? If there is a distinction here, we should add to a brief message regarding how to correctly freeze parameters in the readme.

**Comment by Liyang90** [comment]
> Just want to clarify here, the user should freeze parameters before sharding? 

Right 

> What difference does it make if the parameters are frozen after sharding? If there is a distinction here, we should add to a brief message regarding how to correctly freeze parameters in the readme.
The line of code here:
 https://github.com/pytorch/xla/blob/8a98a88724248fc3ce96a486d57f3dc81a2d260e/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L697
and 
https://github.com/pytorch/xla/blob/8a98a88724248fc3ce96a486d57f3dc81a2d260e/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L707

indicates the original `requires_grad` status is copied over to the sharded copy of the parameter and the full param place holder tensor as well. So it needs to be correctly set by then for the sharded parameter to have the correct value.

Setting `requires_grad` after wrapping my not have the behavior as expected.

The ultimate cause of the issue is the `post_backward_hook` does not fire on params that do not `requires_grad`, so setting `requires_grad=False` after wrapping seems to mask the issue because FSDP still think these params `requires_grad`.

**Comment by carmocca** [comment]
@Liyang90 To be super clear, do you mean that until this PR lands, you recommend that `requires_grad=False` is done after wrapping?

**Comment by Liyang90** [comment]
> @Liyang90 To be super clear, do you mean that until this PR lands, you recommend that `requires_grad=False` is done **after** wrapping?
> 
> And that it's done **before** after it lands?

No, I do not recommend setting `requires_grad=False` after the wrapping. That may lead to correctness error.

**Comment by carmocca** [comment]
I see, but empirically, the memory usage is decreased by doing it after.

May I ask if this PR will land before the 2.1 release is published? This seems to be a critical bug in the finetuning case. Thank you

**Comment by Liyang90** [comment]
> May I ask if this PR will land before the 2.1 release is published? This seems to be a critical bug in the finetuning case. Thank you

I agree. @alanwaketan

**Comment by Liyang90** [comment]
> I see, but empirically, the memory usage is decreased by doing it after.

That might be because the FSDP wrapper still treats all the weights as `requires_grad==True`

File: torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1299,13 +1299,19 @@ def _wait_for_post_backward(self) -> None:
     # A backward pass is done, clean up below.
     def _finalize_parameters(fsdp_module: XlaFullyShardedDataParallel) -> None:
       """Helper used below on all fsdp modules."""
+      frozen_params = []
```
Comment (alanwaketan) [inline torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:1302]: Can we identify which _free_full_params calls doesn't free the full params instead of adding a new pass at the end?

We might have bigger problem there.

Comment (alanwaketan) [reply torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:1302]: Will it be better to do it here?
https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L1327

Comment (alanwaketan) [reply torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:1302]: After looking into the code more, I think this place is fine.

File: test/test_torch_distributed_fsdp_frozen_weight.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import os
+
+os.environ["XLA_DISABLE_FUNCTIONALIZATION"] = "1"
```
Comment (alanwaketan) [inline test/test_torch_distributed_fsdp_frozen_weight.py:3]: May I ask why? Does functionalization prevent this fix from working?

**Comment by alanwaketan** [comment]
Also, can you make a nicer title for the PR before landing?

File: test/test_torch_distributed_fsdp_frozen_weight.py (hunk 2 above)
Comment (Liyang90) [reply test/test_torch_distributed_fsdp_frozen_weight.py:3]: This can be removed probably now. I previously check the `.data` shape of the full param for the test, and that would give me error because https://github.com/pytorch/xla/blob/5dc7e9058806d6f86c4ad7364743a2308ae7b623/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L1430C41-L1430C41 

Now I just test `_has_full_param`, so that's okay.

Comment (alanwaketan) [reply test/test_torch_distributed_fsdp_frozen_weight.py]: .data is deprecated. You shouldn't use that anymore.

**Review by alanwaketan — APPROVED** [review]
LGTM.

**Comment by wonjoolee95** [comment]
@Liyang90, is this ready to be merged? Asking since we want to cherry-pick into the r2.1 branch as well. Thanks!
